### DISCUSSION TITLE
Capturing CDC Events using SQLAlchemy

### DISCUSSION THREAD

**Discussion by zikphil** [sqlalchemy/sqlalchemy#11419] [Usage Questions] [answered]
Hello fantastic community,

I am in need of capturing CDC-style events / audit logs for my API that is using SQLAlchemy. In essence it is very simple, when objects get commited to the DB I want to capture `timestamp now` `user who made the change` `operation (Insert, Update, Delete` `current row/object serialized to json (I have schemas for that already)`.

I have explored the `sqlalchemy-continuum` project but it seems a bit overkill and poorly maintained at the moment. I am looking for maintainer's advices on which event to listen to and what would be the best method to record those changes. So far I was thinking of listening to session `after_flush` and parse what is inside `session.dirty`. Is this still considered the best approach?

Update: I am using PostgreSQL exclusively and I am trying to avoid the complexity of setting up Kafka/Debezium for this.

Thanks!

**Comment by zzzeek** [comment, marked as answer]
I dont know what CDC is (googled, inconclusive), but if you are looking for a simpler version of what sqlalchemy-continuum does, it likely built off our original versioning example which is [here](https://docs.sqlalchemy.org/en/20/orm/examples.html#module-examples.versioned_history).      Your description though seems like you are looking to intercept specific insert/update/delete.   poking around in after_flush is one option, using the mapper level events like [after_insert](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.MapperEvents.after_insert) might be another, you can intercept SQL directly also using before/after execute, before/after cursor execute events.

mostly depends on what you are doing with the observations you make.  if you are looking to add those observations to the database in the same transaction using ORM unit of work, you'd maybe use before_flush instead (that's what the versioning example uses)

**Reply by zikphil** [reply]
Thank you so much. It's crazy all the gems hidden in the documentation, that's enough to get me going.

[https://en.wikipedia.org/wiki/Change_data_capture](https://en.wikipedia.org/wiki/Change_data_capture)

It's essentially just the process of capturing database changes on a row-level for further processing downstream. Or at least that is my understanding of it.
